### ISSUE TITLE
Failure with runGlobalBeforeEachHooks and external class

### ISSUE THREAD

**Issue by wasnotrice** [specta/specta#148]
First, thanks for all your work on Specta. Great library!

I'm getting intermittent failures when running my spec suite with version 0.4 via CocoaPods. The failures involve an external library that isn't involved in the specs, but is part of the app. Error log [here](https://gist.github.com/wasnotrice/6686bcc6144cd87ea0bc).

This external library, `AVYPhotoEditorCustomization` seems to be getting picked up by `[SPTExampleGroup runGlobalBeforeEachHooks:]`. Might it be possible in this method to filter for only Specta-related classes?

I have tried to work around this issue by adding this code to my spec helper (included by all of my spec files), but it doesn't seem to help:

```objc
@interface AVYPhotoEditorCustomization(SPTExclude) <SPTExcludeGlobalBeforeAfterEach>
@end

@implementation AVYPhotoEditorCustomization(SPTExclude)
@end
```

Any guidance for eliminating this failure? Happy to help however I can. Thanks!

**Comment by modocache** [comment]
It does seem unsafe for Specta to pick up *every* class that implements `+beforeEach` and `+afterEach`.  Shouldn't it be *opt-in*, not *opt-out*? That is, *only* classes that conform to a protocol like `SPTGlobalBeforeAfterEach` should be included.

@wasnotrice, can you try putting the following code somewhere it'll be executed before Specta runs its tests, like maybe in your test helper's `+load` method?

```objc
Class klass = NSClassFromString(@"AVYPhotoEditorCustomization");
class_addProtocol(klass, @protocol(SPTExcludeGlobalBeforeAfterEach));
```

**Comment by wasnotrice** [comment]
@modocache that's what I was thinking, too. Thanks for the suggestion. I added that code to my spec helper's `+load` method, but I still get the failure :(

```objc
+ (void)load
{
    [super load];
    Class klass = NSClassFromString(@"AVYPhotoEditorCustomization");
    class_addProtocol(klass, @protocol(SPTExcludeGlobalBeforeAfterEach));
    NSLog(@"Added exclude protocol");
}
```

(I confirmed in my logs that the method did run)

**Comment by paweldudek** [comment]
I'm still leaning towards having this as opt-out rather than opt-in. Here's a gist of my arguments:

* So far we had two issues related to this. I'm pretty sure this issue is unavoidable even with whitelisting mechanism, but more on that in a second. Thus I'm a bit reluctant to seeing this change
* It's a bit more complicated for developers to integrate with this feature (however it does seem likely if that they read about it in docs they'll stumble upon the whitelisting method)
* Requires a bit of work for people who already implement this

Now going back to the problem @wasnotrice has. Let's quickly go through your stack trace:

```
	0   CoreFoundation                      0x0000000107d50f35 __exceptionPreprocess + 165
	1   libobjc.A.dylib                     0x0000000106e56bb7 objc_exception_throw + 45
	2   CoreFoundation                      0x0000000107d50e6d +[NSException raise:format:] + 205
	3   libdispatch.dylib                   0x00000001088bbaf4 _dispatch_client_callout + 8
	4   libdispatch.dylib                   0x00000001088a9e72 dispatch_once_f + 129
	5   popSlate                            0x0000000102d58937 AVYSDKResourceBundle + 312
	6   popSlateTests                       0x0000000113af15d5 +[AVYPhotoEditorCustomization initialize] + 53
	7   libobjc.A.dylib                     0x0000000106e574d6 _class_initialize + 648
	8   libobjc.A.dylib                     0x0000000106e606e1 lookUpImpOrForward + 351
	9   libobjc.A.dylib                     0x0000000106e6d0d3 objc_msgSend + 211
	10  libobjc.A.dylib                     0x0000000106e55599 _class_resolveMethod + 124
	11  libobjc.A.dylib                     0x0000000106e60707 lookUpImpOrForward + 389
	12  libobjc.A.dylib                     0x0000000106e6045f class_getInstanceMethod + 60
	13  popSlateTests                       0x00000001137ec2e5 ClassesWithClassMethod + 245
	14  popSlateTests                       0x00000001137ec1cc __44-[SPTExampleGroup runGlobalBeforeEachHooks:]_block_invoke + 28
	15  libdispatch.dylib                   0x00000001088bbaf4 _dispatch_client_callout + 8
	16  libdispatch.dylib                   0x00000001088a9e72 dispatch_once_f + 129
	17  popSlateTests                       0x00000001137ebf9f -[SPTExampleGroup runGlobalBeforeEachHooks:] + 159
	18  popSlateTests                       0x00000001137ece6f -[SPTExampleGroup runBeforeEachHooks:] + 159
	19  popSlateTests                       0x00000001137ec826 -[SPTExampleGroup runBeforeHooks:] + 118
	20  popSlateTests                       0x00000001137eebd6 __44-[SPTExampleGroup compileExamplesWithStack:]_block_invoke136 + 166
	21  popSlateTests                       0x00000001137f14eb -[SPTSpec spt_runExample:] + 347
	22  popSlateTests                       0x00000001137f0f9d __35+[SPTSpec spt_convertToTestMethod:]_block_invoke + 93
	23  CoreFoundation                      0x0000000107c48c8c __invoking___ + 140
	24  CoreFoundation                      0x0000000107c48ae2 -[NSInvocation invoke] + 290
	25  XCTest                              0x00000001146fe0f7 -[XCTestCase invokeTest] + 253
	26  XCTest                              0x00000001146fe2f8 -[XCTestCase performTest:] + 150
	27  popSlateTests                       0x00000001137f2119 -[SPTSpec performTest:] + 137
	28  XCTest                              0x0000000114707bf5 -[XCTest run] + 260
	29  XCTest                              0x00000001146fcffb -[XCTestSuite performTest:] + 379
	30  XCTest                              0x0000000114707bf5 -[XCTest run] + 260
	31  XCTest                              0x00000001146fcffb -[XCTestSuite performTest:] + 379
	32  XCTest                              0x0000000114707bf5 -[XCTest run] + 260
	33  XCTest                              0x00000001146fcffb -[XCTestSuite performTest:] + 379
	34  XCTest                              0x0000000114707bf5 -[XCTest run] + 260
	35  XCTest                              0x00000001146fa0bc __25-[XCTestDriver _runSuite]_block_invoke + 56
	36  XCTest                              0x0000000114704ead -[XCTestObservationCenter _observeTestExecutionForBlock:] + 162
	37  XCTest                              0x00000001146f9ff0 -[XCTestDriver _runSuite] + 269
	38  XCTest                              0x00000001146faa1d -[XCTestDriver _checkForTestManager] + 234
	39  XCTest                              0x000000011470aaf0 +[XCTestProbe runTests:] + 182
	40  Foundation                          0x0000000106a0fd55 __NSFireDelayedPerform + 387
	41  CoreFoundation                      0x0000000107cb8f64 __CFRUNLOOP_IS_CALLING_OUT_TO_A_TIMER_CALLBACK_FUNCTION__ + 20
	42  CoreFoundation                      0x0000000107cb8b25 __CFRunLoopDoTimer + 1045
	43  CoreFoundation                      0x0000000107c7be5d __CFRunLoopRun + 1901
	44  CoreFoundation                      0x0000000107c7b486 CFRunLoopRunSpecific + 470
	45  GraphicsServices                    0x000000010948d9f0 GSEventRunModal + 161
	46  UIKit                               0x00000001056ff420 UIApplicationMain + 1282
	47  popSlate                            0x0000000102ba3c33 main + 115
	48  libdyld.dylib                       0x00000001088ec145 start + 1
```

The issue here is that there our code results in ` +[AVYPhotoEditorCustomization initialize]` being called. `+[NSObject initialize]` is called only when a class gets loaded into runtime. Docs are a bit more specific:

> The runtime sends initialize to each class in a program just before the class, or any class that inherits from it, is sent its first message from within the program. The runtime sends the initialize message to classes in a thread-safe manner. Superclasses receive this message before their subclasses. The superclass implementation may be called multiple times if subclasses do not implement initialize—the runtime will call the inherited implementation—or if subclasses explicitly call [super initialize]. If you want to protect yourself from being run multiple times, you can structure your implementation along these lines:

> Because initialize is called in a thread-safe manner and the order of initialize being called on different classes is not guaranteed, it’s important to do the minimum amount of work necessary in initialize methods. Specifically, any code that takes locks that might be required by other classes in their initialize methods is liable to lead to deadlocks. Therefore you should not rely on initialize for complex initialization, and should instead limit it to straightforward, class local initialization.

It seems that Aviary is doing some weird stuff in their initialize method which then leads to this crash. 

Now the reason why I think this will not work. Code responsible for both blacklisting and whitelisting is nearly the same:

Blacklisting:

```
class_conformsToProtocol(aClass, @protocol(SPTExcludeGlobalBeforeAfterEach)) == NO
```

Whitelisting:

```
class_conformsToProtocol(aClass, @protocol(SPTIncludeGlobalBeforeAfterEach)) == YES
``` 

and thus I don't see any reason why it will not call `+[NSObject initialize]` as well, which will, unfortunately, result in the same issue as @wasnotrice described. 

The real solution here would be to add logic that somehow checks whether a class conforms to protocol without actually initializing it. I'm not sure how `class_conformsToProtocol` behaves and whether it actually initializes classes. 

Another thing you can try is to avoid `NSClassFromString(@"AVYPhotoEditorCustomization");` and use `objc_lookUpClass` which is a bit more low-level and might actually not call `+[NSObject initialize]`.

**Comment by paweldudek** [comment]
So I made a quick test. I've created a class `Test` with `+[NSObject initialize]` overridden. 

I've put a breakpoint in the aforementioned method and checked whether it was called for specific scenarios:

* Without any additional action `+[NSObject initialize]` was called
* Adding exclude protocol via category for `Test`: `+[NSObject initialize]` was not called
* Adding exclude protocol via `NSClassFromString`: `+[NSObject initialize]` was not called
* Adding exclude protocol via `objc_lookUpClass`: `+[NSObject initialize]` was not called

@wasnotrice could you please check what hits `+[AVYPhotoEditorCustomization initialize]` if it has the exclude protocol added?

**Comment by modocache** [comment]
> @wasnotrice could you please check what hits +[AVYPhotoEditorCustomization initialize] if it has the exclude protocol added?

For the scope of this issue, I think it's worth double-checking here. :point_up: 

But a few notes:

> So far we had two issues related to this. I'm pretty sure this issue is unavoidable even with whitelisting mechanism

There are [other approaches](https://github.com/Quick/Quick/blob/47079aea66854e9398cf211037d7409bf40b2e5d/Documentation/ConfiguringQuick.md#adding-global-beforeeach-and-aftereach-closures), ones that doesn't involve checking the class list for protocol conformance in either direction. I don't think that's the only way to do it, and would love to see more, but I don't necessarily think looking for protocol *conformance* (as opposed to adoption, or some other method) is the best approach. As you mention, there have already been two false positives since the feature was introduced.

> It's a bit more complicated for developers to integrate with this feature

Maybe it's just me, but I struggle to think of when this feature might be necessary. I know of only two users in the real world: @wasnotrice and [libgit2/objective-git](https://github.com/libgit2/objective-git). I'd rather it be a little harder to use than for users to accidentally have it fire on them. I mean, take a look at [the PR that introduced the opt-out feature](https://github.com/specta/specta/pull/131/files): it includes a code comment that says "If you're seeing a crash here then...", which I think is a pretty bad experience for users.

> Requires a bit of work for people who already implement this

This, I think, is the biggest problem with changing how it currently works. I don't want to change an API that people rely upon. https://github.com/specta/specta/pull/149 does that, and users who use the feature are going to have to make changes to keep up. I think it's worth it to make sure that's absolutely necessary in this case, and whether they may be a better approach to global before/after each as a whole.

**Comment by orta** [comment]
I think we should break it, and call the next release `1.0.0` with whitelisting. Feels like a weird feature in general to me.

**Comment by modocache** [comment]
That sounds fine to me. I think whitelisting sounds pretty safe.

> I'm pretty sure this issue is unavoidable even with whitelisting mechanism

@paweldudek Can you explain your thoughts a bit? From what I can tell, the problem here is that @wasnotrice is trying to opt-out *before* his class is picked up by Specta's code, so he needed to find a way to dynamically opt-out in time. With a whitelisting mechanism, users won't accidentally be opted-in, so they won't have to race Specta in order to prevent a crash. :snail:

**Comment by paweldudek** [comment]
@modocache no problem! 

So the issue @wasnotrice is seeing is caused by Aviary SDK hitting an assertion in their `[NSObject initialize]` method (which sounds like a horrible design, but that's a different story :wink:). Specta tries to send `beforeEach` (as we can see in the stack trace) message to that class, which triggers `initialize`, which then results in an exception. 

`initialize` method is, according to docs, sent before first message is received by given class. And is sent only once per given class (however since some classes do not implement it superclasses can receive this call multiple times). 

@wasnotrice noted that he tried to exclude that class using previous method with blacklisting protocol. However that did not work and the exception was still raised. 

Now there are two possibilities here. Either previous blacklisting mechanism was broken (as it allowed that class to somehow slip through, in this case I'm 100% for changing to whitelisting) or something else triggered that `[NSObject initialize]` call (that's why I'd love to see the stack trace for this exception when blacklisting is added). 

It is possible that code for adding blacklisting protocol somehow resulted in an `initialize` message being sent to `AVYPhotoEditorCustomization`, but I'm not sure how is that possible as my simple test confirmed that using

* `objc_lookUpClass`
* `NSClassFromString`

does not trigger `initialize`. 

And the opt-out mechanism works at the same time as opt-in mechanism so, from behavior perspective, there should be no difference between adding a blacklisting/whitelisting here (aka the `ClassesWithClassMethod` function). 

Also, I sort of convinced myself that whitelisting might be a better idea - it's way more complicated to explain what to do if you don't want to use global before/after each methods rather than explain what to do if you want to have them. 

> Maybe it's just me, but I struggle to think of when this feature might be necessary.

I might be opposing this a bit as I know how many global hooks like that we have in our codebase and my subconscious tells me it's gonna be quite a lot of work to update all our private pods to make specs work again :wink:

**Comment by wasnotrice** [comment]
Thanks all for your feedback and consideration.

@paweldudek The [failure is the same](https://gist.github.com/wasnotrice/2bcb6ab2dcde4a8b8967) when I add the blacklist protocol, whether I use `NSClassFromString()` or `objc_lookUpClass()`. The log shows that the protocol was added to the class, and those are the very first lines of the log, so it seems like it's happening as early as possible.

This is an intermittent failure. Tests tend to pass once or twice and then start failing. Once they start failing, it seems like they fail more frequently, but that may be psychological. In any case, I get this error at least 30% of the time.

Since the failure is intermittent, I can't be sure that the whitelisting method actually prevents it, but I have run the suite 15 times or so without error, which never happens with blacklisting.

@orta I don't know if you have other reasons for moving to 1.0, but I don't think there's anything wrong with breaking changes in a pre-1.0 project. I'm prepared for that when I use a project at 0.4.0 :)

I agree with @paweldudek that the whitelisting approach is easier to explain/understand. I was surprised that I had to add code to _prevent_ Specta from executing methods in my codebase that I didn't want to execute.

**Comment by orta** [comment]
The 1.0.0 is an ideological thing, Specta is very production worthy. Its version should reflect that.

**Comment by wasnotrice** [comment]
:+1:

**Comment by wasnotrice** [comment]
Fixed by #149. Thanks everyone!

**Comment by modocache** [comment]
Thank *you*! :100:
